src/gamepad/mapping.vala

```diff
@@ -0,0 +1,114 @@
+// This file is part of GNOME Games. License: GPLv3
+
+private class LibGamepad.Dpad : Object {
+	public InputType types[4];
+	public int values[4];
+	public int axisval[2];
+
+	public Dpad () {
+		axisval[0] = axisval[1] = 0;
+		types[0] = types[1] = types[2] = types[3] = InputType.INVALID;
+	}
+}
+
+public class LibGamepad.Mapping : Object {
+	private InputType[] buttons_type;
+	private int[] buttons_value;
+	private InputType[] axes_type;
+	private int[] axes_value;
+	private Dpad[] dpads;
+
+	public Mapping.from_sdl_string (string? mappingstring) throws MappingError {
+		if (mappingstring == null || mappingstring == "")
+			throw new MappingError.NOT_A_MAPPING ("mappingstring cannot be null or empty");
```

Comment:
Make it a sentence, something like "The mapping string can't be null nor empty."